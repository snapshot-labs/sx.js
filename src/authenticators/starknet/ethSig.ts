import { Call, CallData, shortString } from 'starknet';
import EthSigAuthenticatorAbi from './abis/EthSigAuthenticator.json';
import { getChoiceEnum } from '../../utils/starknet-enums';
import {
  Authenticator,
  Envelope,
  Propose,
  Vote,
  UpdateProposal,
  ProposeCallArgs,
  VoteCallArgs,
  UpdateProposalCallArgs
} from '../../types';

const callData = new CallData(EthSigAuthenticatorAbi);

export default function createEthSigAuthenticator(): Authenticator {
  return {
    type: 'ethSig',
    createProposeCall(envelope: Envelope<Propose>, args: ProposeCallArgs): Call {
      const { space, authenticator } = envelope.data;

      if (!envelope.signatureData?.signature) {
        throw new Error('signature is required for this authenticator');
      }

      if (!envelope.signatureData?.message) {
        throw new Error('message is required for this authenticator');
      }

      const [r, s, v] = envelope.signatureData.signature;

      const compiled = callData.compile('authenticate_propose', [
        r,
        s,
        v,
        space,
        args.author,
        shortString.splitLongString(args.metadataUri),
        {
          address: args.executionStrategy.address,
          params: args.executionStrategy.params
        },
        args.strategiesParams,
        envelope.signatureData.message.salt
      ]);

      return {
        contractAddress: authenticator,
        entrypoint: 'authenticate_propose',
        calldata: compiled
      };
    },
    createVoteCall(envelope: Envelope<Vote>, args: VoteCallArgs): Call {
      const { space, authenticator } = envelope.data;

      if (!envelope.signatureData?.signature) {
        throw new Error('signature is required for this authenticator');
      }

      if (!envelope.signatureData?.message) {
        throw new Error('message is required for this authenticator');
      }

      const [r, s, v] = envelope.signatureData.signature;

      const compiled = callData.compile('authenticate_vote', [
        r,
        s,
        v,
        space,
        args.voter,
        args.proposalId,
        getChoiceEnum(args.choice),
        args.votingStrategies.map(strategy => ({
          index: strategy.index,
          params: strategy.params
        })),
        shortString.splitLongString(args.metadataUri)
      ]);

      return {
        contractAddress: authenticator,
        entrypoint: 'authenticate_vote',
        calldata: compiled
      };
    },
    createUpdateProposalCall(
      envelope: Envelope<UpdateProposal>,
      args: UpdateProposalCallArgs
    ): Call {
      const { space, authenticator } = envelope.data;

      if (!envelope.signatureData?.signature) {
        throw new Error('signature is required for this authenticator');
      }

      if (!envelope.signatureData?.message) {
        throw new Error('message is required for this authenticator');
      }

      const [r, s, v] = envelope.signatureData.signature;

      const compiled = callData.compile('authenticate_update_proposal', [
        r,
        s,
        v,
        space,
        {
          address: args.author
        },
        args.proposalId,
        {
          address: args.executionStrategy.address,
          params: args.executionStrategy.params
        },
        shortString.splitLongString(args.metadataUri),
        envelope.signatureData.message.salt
      ]);

      return {
        contractAddress: authenticator,
        entrypoint: 'authenticate_update_proposal',
        calldata: compiled
      };
    }
  };
}
