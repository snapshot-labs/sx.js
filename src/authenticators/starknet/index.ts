import createVanillaAuthenticator from './vanilla';
import createEthSigAuthenticator from './ethSig';
import createEthTxAuthenticator from './ethTx';
import createStarkSigAuthenticator from './starkSig';
import createStarkTxAuthenticator from './starkTx';
import { hexPadLeft } from '../../utils/encoding';
import type { Authenticator, NetworkConfig } from '../../types';

export function getAuthenticator(
  address: string,
  networkConfig: NetworkConfig
): Authenticator | null {
  const authenticator = networkConfig.authenticators[hexPadLeft(address)];
  if (!authenticator) return null;

  if (authenticator.type === 'vanilla') {
    return createVanillaAuthenticator();
  }

  if (authenticator.type === 'ethSig') {
    return createEthSigAuthenticator();
  }

  if (authenticator.type === 'ethTx') {
    return createEthTxAuthenticator();
  }

  if (authenticator.type === 'starkSig') {
    return createStarkSigAuthenticator();
  }

  if (authenticator.type === 'starkTx') {
    return createStarkTxAuthenticator();
  }

  return null;
}
