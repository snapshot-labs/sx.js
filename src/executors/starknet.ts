import { createStarknetExecutionParams } from '../utils/encoding';
import type { Call } from 'starknet';

export default function createStarknetExecutor() {
  return {
    type: 'starknet',
    getExecutionData(executorAddress: string, calls: Call[]) {
      return {
        executor: executorAddress,
        executionParams: createStarknetExecutionParams(calls)
      };
    }
  };
}
