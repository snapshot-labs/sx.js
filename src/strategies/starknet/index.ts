import createVanillaStrategy from './vanilla';
import createMerkleWhitelistStrategy from './merkleWhitelist';
import createErc20VotesStrategy from './erc20Votes';
import { hexPadLeft } from '../../utils/encoding';
import type { Strategy, NetworkConfig } from '../../types';

export function getStrategy(address: string, networkConfig: NetworkConfig): Strategy | null {
  const strategy = networkConfig.strategies[hexPadLeft(address)];
  if (!strategy) return null;

  if (strategy.type === 'vanilla') {
    return createVanillaStrategy();
  }

  if (strategy.type === 'whitelist') {
    return createMerkleWhitelistStrategy();
  }

  if (strategy.type === 'erc20Votes') {
    return createErc20VotesStrategy();
  }

  return null;
}
