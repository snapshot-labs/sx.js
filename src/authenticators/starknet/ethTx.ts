import { Call, CallData } from 'starknet';
import EthTxAuthenticatorAbi from './abis/EthTxAuthenticator.json';
import {
  Authenticator,
  Envelope,
  Propose,
  Vote,
  UpdateProposal,
  ProposeCallArgs,
  VoteCallArgs,
  UpdateProposalCallArgs
} from '../../types';

const callData = new CallData(EthTxAuthenticatorAbi);

export default function createEthTxAuthenticator(): Authenticator {
  return {
    type: 'ethTx',
    createProposeCall(envelope: Envelope<Propose>, args: ProposeCallArgs): Call {
      const { space, authenticator } = envelope.data.message;

      const compiled = callData.compile('authenticate_propose', [
        space,
        args.author,
        {
          address: args.executionStrategy.address,
          params: [args.executionStrategy.params]
        },
        args.strategiesParams
      ]);

      return {
        contractAddress: authenticator,
        entrypoint: 'authenticate_propose',
        calldata: compiled
      };
    },
    createVoteCall(envelope: Envelope<Vote>, args: VoteCallArgs): Call {
      const { space, authenticator } = envelope.data.message;

      const compiled = callData.compile('authenticate_vote', [
        space,
        args.voter,
        args.proposalId,
        args.choice,
        args.votingStrategies.map(strategy => ({
          index: strategy.index,
          params: [strategy.params]
        }))
      ]);

      return {
        contractAddress: authenticator,
        entrypoint: 'authenticate_vote',
        calldata: compiled
      };
    },
    createUpdateProposalCall(
      envelope: Envelope<UpdateProposal>,
      args: UpdateProposalCallArgs
    ): Call {
      const { space, authenticator } = envelope.data.message;

      const compiled = callData.compile('authenticate_update_proposal', [
        space,
        args.author,
        args.proposalId,
        {
          address: args.executionStrategy.address,
          params: [args.executionStrategy.params]
        }
      ]);

      return {
        contractAddress: authenticator,
        entrypoint: 'authenticate_update_proposal',
        calldata: compiled
      };
    }
  };
}
