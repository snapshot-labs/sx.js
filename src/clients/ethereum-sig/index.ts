import { hash, defaultProvider } from 'starknet';
import randomBytes from 'randombytes';
import { Web3Provider } from '@ethersproject/providers';
import { Wallet } from '@ethersproject/wallet';
import fetch from 'cross-fetch';
import { getStrategy } from '../../strategies';
import { domain, proposeTypes, voteTypes } from './types';
import * as utils from '../../utils';
import {
  Propose,
  Vote,
  EthSigProposeMessage,
  EthSigVoteMessage,
  Envelope,
  EthereumSigClientConfig
} from '../../types';

export class EthereumSig {
  config: EthereumSigClientConfig;

  constructor(config: EthereumSigClientConfig) {
    this.config = config;
  }

  generateSalt() {
    return Number(
      utils.splitUint256.SplitUint256.fromHex(utils.bytes.bytesToHex(randomBytes(4))).toHex()
    );
  }

  async getStrategiesAddresses(data: Propose | Vote) {
    return Promise.all(
      data.strategies.map(
        (id) =>
          defaultProvider.getStorageAt(
            data.space,
            utils.encoding.getStorageVarAddress('Voting_voting_strategies_store', id.toString(16))
          ) as Promise<string>
      )
    );
  }

  async getStrategiesParams(call: 'propose' | 'vote', address: string, data: Propose | Vote) {
    const strategiesAddresses = await this.getStrategiesAddresses(data);

    return Promise.all(
      strategiesAddresses.map((strategyAddress, index) => {
        const strategy = getStrategy(strategyAddress);
        if (!strategy) throw new Error('Invalid strategy');

        return strategy.getParams(
          call,
          strategyAddress,
          index,
          {
            address,
            sig: '',
            data: {
              message: data
            }
          },
          this.config
        );
      })
    );
  }

  public async sign<T extends EthSigProposeMessage | EthSigVoteMessage>(
    web3: Web3Provider | Wallet,
    address: string,
    message: T,
    types
  ): Promise<Envelope<T>> {
    const signer = Wallet.isSigner(web3) ? web3 : web3.getSigner();
    const data = { domain, types, message };
    const sig = await signer._signTypedData(domain, data.types, message);
    return { address, sig, data } as any;
  }

  public async send(envelop) {
    const init = {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        jsonrpc: '2.0',
        method: 'send',
        params: { envelop },
        id: null
      })
    };
    const res = await fetch(this.config.manaUrl, init);
    const json = await res.json();
    return json.result;
  }

  public async propose(web3: Web3Provider | Wallet, address: string, data: Propose) {
    const strategiesParams = await this.getStrategiesParams('propose', address, data);

    const message: EthSigProposeMessage = {
      ...data,
      space: utils.encoding.hexPadRight(data.space),
      authenticator: utils.encoding.hexPadRight(data.authenticator),
      proposerAddress: utils.encoding.hexPadRight(address),
      executor: utils.encoding.hexPadRight(data.executor),
      executionParamsHash: utils.encoding.hexPadRight(
        hash.computeHashOnElements(data.executionParams)
      ),
      usedVotingStrategiesHash: utils.encoding.hexPadRight(
        hash.computeHashOnElements(data.strategies.map((strategy) => `0x${strategy.toString(16)}`))
      ),
      userVotingStrategyParamsFlatHash: utils.encoding.hexPadRight(
        hash.computeHashOnElements(utils.encoding.flatten2DArray(strategiesParams))
      ),
      salt: this.generateSalt()
    };

    return this.sign(web3, address, message, proposeTypes);
  }

  public async vote(web3: Web3Provider | Wallet, address: string, data: Vote) {
    const strategiesParams = await this.getStrategiesParams('vote', address, data);

    const message: EthSigVoteMessage = {
      ...data,
      space: utils.encoding.hexPadRight(data.space),
      authenticator: utils.encoding.hexPadRight(data.authenticator),
      voterAddress: utils.encoding.hexPadRight(address),
      usedVotingStrategiesHash: utils.encoding.hexPadRight(
        hash.computeHashOnElements(data.strategies.map((strategy) => `0x${strategy.toString(16)}`))
      ),
      userVotingStrategyParamsFlatHash: utils.encoding.hexPadRight(
        hash.computeHashOnElements(utils.encoding.flatten2DArray(strategiesParams))
      ),
      salt: this.generateSalt()
    };

    return this.sign(web3, address, message, voteTypes);
  }
}
