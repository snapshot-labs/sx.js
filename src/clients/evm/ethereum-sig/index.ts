import randomBytes from 'randombytes';
import { AbiCoder } from '@ethersproject/abi';
import { SplitUint256 } from '../../../utils/split-uint256';
import { bytesToHex } from '../../../utils/bytes';
import { getStrategiesWithParams } from '../../../strategies/evm';
import { evmGoerli } from '../../../networks';
import { domain as baseDomain, proposeTypes, updateProposalTypes, voteTypes } from './types';
import type { Signer, TypedDataSigner, TypedDataField } from '@ethersproject/abstract-signer';
import type {
  Propose,
  UpdateProposal,
  Vote,
  Envelope,
  SignatureData,
  EIP712ProposeMessage,
  EIP712UpdateProposalMessage,
  EIP712VoteMessage
} from '../types';
import type { EvmNetworkConfig } from '../../../types';

type EthereumSigClientOpts = {
  networkConfig?: EvmNetworkConfig;
  manaUrl?: string;
};

export class EthereumSig {
  manaUrl: string;
  networkConfig: EvmNetworkConfig;

  constructor(opts?: EthereumSigClientOpts) {
    this.networkConfig = opts?.networkConfig || evmGoerli;
    this.manaUrl = opts?.manaUrl || 'https://mana.pizza';
  }

  generateSalt() {
    return Number(SplitUint256.fromHex(bytesToHex(randomBytes(4))).toHex());
  }

  public async sign<
    T extends EIP712ProposeMessage | EIP712UpdateProposalMessage | EIP712VoteMessage
  >(
    signer: Signer & TypedDataSigner,
    verifyingContract: string,
    message: T,
    types: Record<string, TypedDataField[]>
  ): Promise<SignatureData> {
    const address = await signer.getAddress();

    const domain = {
      ...baseDomain,
      chainId: this.networkConfig.eip712ChainId,
      verifyingContract
    };

    const signature = await signer._signTypedData(domain, types, message);

    return {
      address,
      signature,
      domain,
      types,
      message
    };
  }

  public async send(envelope) {
    const body = {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        jsonrpc: '2.0',
        method: 'send',
        params: { envelope },
        id: null
      })
    };

    const res = await fetch(`${this.manaUrl}/eth_rpc/${this.networkConfig.eip712ChainId}`, body);
    const json = await res.json();

    return json.result;
  }

  public async propose({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: Propose;
  }): Promise<Envelope<Propose>> {
    const author = await signer.getAddress();

    const userStrategies = await getStrategiesWithParams(
      'propose',
      data.strategies,
      author,
      data,
      this.networkConfig
    );

    const abiCoder = new AbiCoder();
    const message: EIP712ProposeMessage = {
      space: data.space,
      author,
      metadataURI: data.metadataUri,
      executionStrategy: data.executionStrategy,
      userProposalValidationParams: abiCoder.encode(
        ['tuple(int8 index, bytes params)[]'],
        [userStrategies]
      ),
      salt: this.generateSalt()
    };

    const signatureData = await this.sign(signer, data.authenticator, message, proposeTypes);

    return {
      signatureData,
      data
    };
  }

  public async updateProposal({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: UpdateProposal;
  }): Promise<Envelope<UpdateProposal>> {
    const author = await signer.getAddress();

    const message: EIP712UpdateProposalMessage = {
      space: data.space,
      author,
      proposalId: data.proposal,
      executionStrategy: data.executionStrategy,
      metadataURI: data.metadataUri,
      salt: this.generateSalt()
    };

    const signatureData = await this.sign(signer, data.authenticator, message, updateProposalTypes);

    return {
      signatureData,
      data
    };
  }

  public async vote({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: Vote;
  }): Promise<Envelope<Vote>> {
    const voter = await signer.getAddress();

    const userVotingStrategies = await getStrategiesWithParams(
      'vote',
      data.strategies,
      voter,
      data,
      this.networkConfig
    );

    const message: EIP712VoteMessage = {
      space: data.space,
      voter,
      proposalId: data.proposal,
      choice: data.choice,
      userVotingStrategies,
      voteMetadataURI: data.metadataUri
    };

    const signatureData = await this.sign(signer, data.authenticator, message, voteTypes);

    return {
      signatureData,
      data
    };
  }
}
