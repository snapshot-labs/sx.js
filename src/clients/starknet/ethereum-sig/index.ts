import randomBytes from 'randombytes';
import { Signer, TypedDataSigner, TypedDataField } from '@ethersproject/abstract-signer';
import { CallData, shortString } from 'starknet';
import { getStrategiesWithParams } from '../../../utils/strategies';
import { proposeTypes, updateProposalTypes, voteTypes } from './types';
import {
  ClientConfig,
  ClientOpts,
  Envelope,
  Propose,
  UpdateProposal,
  Vote,
  EIP712ProposeMessage,
  EIP712UpdateProposalMessage,
  EIP712VoteMessage,
  SignatureData
} from '../../../types';
import { defaultNetwork } from '../../..';
import { getRSVFromSig } from '../../../utils/encoding';

export class EthereumSig {
  config: ClientConfig;

  constructor(opts: ClientOpts) {
    this.config = {
      networkConfig: defaultNetwork,
      ...opts
    };
  }

  generateSalt() {
    return `0x${randomBytes(4).toString('hex')}`;
  }

  public async sign<
    T extends EIP712ProposeMessage | EIP712UpdateProposalMessage | EIP712VoteMessage
  >(
    signer: Signer & TypedDataSigner,
    message: T,
    types: Record<string, TypedDataField[]>,
    primaryType: string
  ): Promise<SignatureData> {
    const address = await signer.getAddress();

    const domain = {};

    const extendedMessage = {
      chainId: this.config.networkConfig.eip712ChainId,
      ...message
    };

    const signature = await signer._signTypedData(domain, types, extendedMessage);
    const { r, s, v } = getRSVFromSig(signature);

    return {
      address,
      signature: [r.toHex(), s.toHex(), v],
      message: extendedMessage,
      primaryType
    };
  }

  public async propose({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: Propose;
  }): Promise<Envelope<Propose>> {
    const address = await signer.getAddress();

    const userStrategies = await getStrategiesWithParams(
      'propose',
      data.strategies,
      address,
      data,
      this.config
    );

    const message = {
      authenticator: data.authenticator,
      space: data.space,
      author: address,
      executionStrategy: {
        address: data.executionStrategy.addr,
        params: data.executionStrategy.params
      },
      userProposalValidationParams: CallData.compile({
        userStrategies
      }),
      metadataUri: shortString
        .splitLongString(data.metadataUri)
        .map(str => shortString.encodeShortString(str)),
      salt: this.generateSalt()
    };

    const signatureData = await this.sign(signer, message, proposeTypes, 'Propose');

    return {
      signatureData,
      data
    };
  }

  public async updateProposal({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: UpdateProposal;
  }): Promise<Envelope<UpdateProposal>> {
    const address = await signer.getAddress();

    const message = {
      authenticator: data.authenticator,
      space: data.space,
      author: address,
      proposalId: `0x${data.proposal.toString(16)}`,
      executionStrategy: {
        address: data.executionStrategy.addr,
        params: data.executionStrategy.params
      },
      metadataUri: shortString
        .splitLongString(data.metadataUri)
        .map(str => shortString.encodeShortString(str)),
      salt: this.generateSalt()
    };

    const signatureData = await this.sign(signer, message, updateProposalTypes, 'UpdateProposal');

    return {
      signatureData,
      data
    };
  }

  public async vote({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: Vote;
  }): Promise<Envelope<Vote>> {
    const address = await signer.getAddress();

    const userVotingStrategies = await getStrategiesWithParams(
      'vote',
      data.strategies,
      address,
      data,
      this.config
    );

    const message = {
      authenticator: data.authenticator,
      space: data.space,
      voter: address,
      proposalId: `0x${data.proposal.toString(16)}`,
      choice: `0x${data.choice.toString(16)}`,
      userVotingStrategies,
      metadataUri: shortString.splitLongString('').map(str => shortString.encodeShortString(str))
    };

    const signatureData = await this.sign(signer, message, voteTypes, 'Vote');

    return {
      signatureData,
      data
    };
  }
}
