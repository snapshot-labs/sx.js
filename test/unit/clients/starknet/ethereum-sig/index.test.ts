import { Provider } from 'starknet';
import { JsonRpcProvider } from '@ethersproject/providers';
import { Wallet } from '@ethersproject/wallet';
import { EthereumSig } from '../../../../../src/clients/starknet/ethereum-sig';

describe('EthereumSig', () => {
  const starkProvider = new Provider({
    sequencer: {
      baseUrl: 'http://127.0.0.1:5050'
    }
  });

  const provider = new JsonRpcProvider('https://rpc.brovider.xyz/5');
  const signer = new Wallet(
    '0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80',
    provider
  );

  const client = new EthereumSig({
    starkProvider,
    ethUrl: 'https://rpc.brovider.xyz/5'
  });

  beforeAll(() => {
    jest.spyOn(client, 'generateSalt').mockImplementation(() => '0x0');
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should create propose envelope', async () => {
    const envelope = await client.propose({
      signer,
      data: {
        space: '0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8',
        authenticator: '0x00e6533da3322019c3e26bd6942b647a74593af805021003bab707267717952a',
        strategies: [
          {
            address: '0x510d1e6d386a2adcfc6f2a57f80c4c4268baeccbd4a09334e843b17ce9225ee',
            index: 0
          }
        ],
        executionStrategy: {
          addr: '0x040de235a2b53e921d37c2ea2b160750ca2e94f01d709f78f870963559de8fbe',
          params: ['0x101']
        },
        metadataUri: 'ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca'
      }
    });

    expect(envelope).toMatchSnapshot();
  });

  it('should create update proposal envelope', async () => {
    const envelope = await client.updateProposal({
      signer,
      data: {
        space: '0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8',
        authenticator: '0x00e6533da3322019c3e26bd6942b647a74593af805021003bab707267717952a',
        executionStrategy: {
          addr: '0x040de235a2b53e921d37c2ea2b160750ca2e94f01d709f78f870963559de8fbe',
          params: ['0x101']
        },
        proposal: 1,
        metadataUri: 'ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca'
      }
    });

    expect(envelope).toMatchSnapshot();
  });

  it('should create vote envelope', async () => {
    const envelope = await client.vote({
      signer,
      data: {
        space: '0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8',
        authenticator: '0x00e6533da3322019c3e26bd6942b647a74593af805021003bab707267717952a',
        strategies: [
          {
            address: '0x510d1e6d386a2adcfc6f2a57f80c4c4268baeccbd4a09334e843b17ce9225ee',
            index: 0
          }
        ],
        proposal: 1,
        choice: 1
      }
    });

    expect(envelope).toMatchSnapshot();
  });
});
