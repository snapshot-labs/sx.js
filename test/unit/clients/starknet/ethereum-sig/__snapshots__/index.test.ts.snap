// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EthereumSig should create propose envelope 1`] = `
{
  "data": {
    "authenticator": "0x00e6533da3322019c3e26bd6942b647a74593af805021003bab707267717952a",
    "executionStrategy": {
      "addr": "0x040de235a2b53e921d37c2ea2b160750ca2e94f01d709f78f870963559de8fbe",
      "params": [
        "0x101",
      ],
    },
    "metadataUri": "ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca",
    "space": "0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8",
    "strategies": [
      {
        "address": "0x510d1e6d386a2adcfc6f2a57f80c4c4268baeccbd4a09334e843b17ce9225ee",
        "index": 0,
      },
    ],
  },
  "signatureData": {
    "address": "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
    "message": {
      "authenticator": "0x00e6533da3322019c3e26bd6942b647a74593af805021003bab707267717952a",
      "author": "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
      "chainId": "0x534e5f474f45524c49",
      "executionStrategy": {
        "address": "0x040de235a2b53e921d37c2ea2b160750ca2e94f01d709f78f870963559de8fbe",
        "params": [
          "0x101",
        ],
      },
      "metadataUri": [
        "0x697066733a2f2f516d4e726d36784b75696231544874576b694e35434b7442",
        "0x4565725143447055746d6744716961553278446d6361",
      ],
      "salt": "0x0",
      "space": "0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8",
      "userProposalValidationParams": [
        "1",
        "0",
        "1",
        "0",
      ],
    },
    "primaryType": "Propose",
    "signature": [
      "0x2f3f31908e99a8a6c37b14cee6cc334096adeae8b327442a0d4fd663ef7221",
      "0x6d7cb169150067bd1f1b020c4987a3b903ef796d28b6916e4e284401ea381750",
      "0x1c",
    ],
  },
}
`;

exports[`EthereumSig should create update proposal envelope 1`] = `
{
  "data": {
    "authenticator": "0x00e6533da3322019c3e26bd6942b647a74593af805021003bab707267717952a",
    "executionStrategy": {
      "addr": "0x040de235a2b53e921d37c2ea2b160750ca2e94f01d709f78f870963559de8fbe",
      "params": [
        "0x101",
      ],
    },
    "metadataUri": "ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca",
    "proposal": 1,
    "space": "0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8",
  },
  "signatureData": {
    "address": "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
    "message": {
      "authenticator": "0x00e6533da3322019c3e26bd6942b647a74593af805021003bab707267717952a",
      "author": "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
      "chainId": "0x534e5f474f45524c49",
      "executionStrategy": {
        "address": "0x040de235a2b53e921d37c2ea2b160750ca2e94f01d709f78f870963559de8fbe",
        "params": [
          "0x101",
        ],
      },
      "metadataUri": [
        "0x697066733a2f2f516d4e726d36784b75696231544874576b694e35434b7442",
        "0x4565725143447055746d6744716961553278446d6361",
      ],
      "proposalId": "0x1",
      "salt": "0x0",
      "space": "0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8",
    },
    "primaryType": "UpdateProposal",
    "signature": [
      "0x43e87e09c3cefce1e65fc6d138eddcb390c70cd8f7610eef2f5f702e8a51f1a8",
      "0x26c2790340ebe733c2b41bb8b6ce98ab78169dff5ee8a57725e25db1219e691d",
      "0x1b",
    ],
  },
}
`;

exports[`EthereumSig should create vote envelope 1`] = `
{
  "data": {
    "authenticator": "0x00e6533da3322019c3e26bd6942b647a74593af805021003bab707267717952a",
    "choice": 1,
    "proposal": 1,
    "space": "0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8",
    "strategies": [
      {
        "address": "0x510d1e6d386a2adcfc6f2a57f80c4c4268baeccbd4a09334e843b17ce9225ee",
        "index": 0,
      },
    ],
  },
  "signatureData": {
    "address": "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
    "message": {
      "authenticator": "0x00e6533da3322019c3e26bd6942b647a74593af805021003bab707267717952a",
      "chainId": "0x534e5f474f45524c49",
      "choice": "0x1",
      "metadataUri": [],
      "proposalId": "0x1",
      "space": "0x06330d3e48f59f5411c201ee2e9e9ccdc738fb3bb192b0e77e4eda26fa1a22f8",
      "userVotingStrategies": [
        {
          "index": 0,
          "params": [
            "0x0",
          ],
        },
      ],
      "voter": "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
    },
    "primaryType": "Vote",
    "signature": [
      "0x837fe6824e2f2ded7bd811b0db06d02244f4f8255d177bff7e0d4ef0f188b689",
      "0x3c04f852c3a8ff50f4840883691192f19f1e32531d5578e558e734740b9c1e25",
      "0x1b",
    ],
  },
}
`;
